src/hotspot/share/cds/filemap.cpp

```diff
@@ -2094,14 +2091,15 @@ bool FileMapInfo::map_heap_region() {
 
     // Make sure we map at the very top of the heap - see comments in
     // init_heap_region_relocation().
-    MemRegion heap_range = G1CollectedHeap::heap()->reserved();
+    MemRegion heap_range = Universe::heap()->reserved();
     assert(heap_range.contains(_mapped_heap_memregion), "must be");
 
-    address heap_end = (address)heap_range.end();
-    address mapped_heap_region_end = (address)_mapped_heap_memregion.end();
-    assert(heap_end >= mapped_heap_region_end, "must be");
-    assert(heap_end - mapped_heap_region_end < (intx)(HeapRegion::GrainBytes),
-           "must be at the top of the heap to avoid fragmentation");
+    if (UseG1GC) {
```

Comment:
This condition is odd. We only call FileMapInfo::map_heap_region() for G1, right? Maybe just assert at the start of the function for `ArchiveHeapLoader::can_map()` instead?

Reply:
On second thought I think there is no need for `ArchiveHeapLoader::can_map()` either. `map_heap_region()` gets called from `FileMapInfo::map_or_load_heap_region` only if `ArchiveHeapLoader::can_map()` is true.